### PR TITLE
DRAFT: DBZ-121 Added options for how change events are mapped to topics

### PR DESCRIPTION
**Pull request by rhauch**
The default is to map all change events in each table to a separate topic, but this commit adds the ability to map all change events for all database tables to a database-specific topic, or to use a custom mapping via `TopicMappingProvider` implementation.

NOTE: This is still incomplete and not thoroughly tested.

### CODE REVIEW

**Comment by hchiorean** [comment]
@rhauch: I've also added a `TOPIC_SELECTION_STRATEGY` config option for the PG connector with the options of `PER_TABLE(default)` or `PER_SCHEMA`. This PR contains the missing pieces in `TableSchemaBuilder` to support this feature (i.e. correct event Key schema and value generator) so I will not attempt to make the same changes to `TableSchemaBuilder`, but rather rely on this once it's merged.

**Comment by rhauch** [comment]
@hchiorean Awesome. Any preferences or suggested changes to names of anything? I do like `TOPIC_SELECTION_STRATEGY` better than `TOPIC_MAPPING`.

**Comment by hchiorean** [comment]
the only comment I would have based on the current proposed PR changes is that I'm not convinced of the benefit of the added complexity for the custom `TopicMappingProvider` option. For PG, there are only 2 options which make sense : per table and per schema. 

As far as integrating with the rest of the code, it's hard to say until I actually get around to doing that, so I'm fine with it in its current form.

**Comment by rhauch** [comment]
@hchiorean, agreed. Ideally we can completely hide how this is implemented, and instead use a configuration option that specifies something like `per-table`, `per-database`, `per-schemas` or whatever reads well. We can expand these if/when we add new built-in strategies.

[DBZ-121](https://issues.jboss.org/browse/DBZ-121) not only deals with expanding how the connector maps events to topics beyond just _topic-per-table_ to also handle _topic-per-database_ and _topic-per-schema_ (whichever makes sense for the DBMS), but it also calls for a way to deal with sharded tables. Originally, I was thinking that dealing with shards is really just a different topic mapping strategy, since you may want to annotate the key to do this. There are so many different approaches to sharding that I'm not sure we can have a built-in approach that works for everyone. Thus the need for specifying a custom mapping strategy.

But just because we could use a custom mapping strategy doesn't mean the configuration should require an implementation class name. This still needs to be changed/refined.

File: debezium-core/src/main/java/io/debezium/relational/topic/ByDatabaseTopicMapping.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            return keySchemaName;
+        }
+        
+        @Override
+        public String getValueSchemaName() {
+            return valueSchemaName;
+        }
+        
+        @Override
+        public void enhanceKeySchema(SchemaBuilder keySchemaBuilder) {
```
Comment (dasl-) [inline debezium-core/src/main/java/io/debezium/relational/topic/ByDatabaseTopicMapping.java]: might be nice to add docs about why this enhancement is necessary.

File: debezium-core/src/main/java/io/debezium/relational/topic/TopicMappingProvider.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+         * @return the topic name; may be null if this strategy could not be applied
+         */
+        String getTopicName();
+
+        /**
+         * Get the name of the schema for the key.
+         * 
+         * @return the key's schema name; may not be null
+         */
+        default String getKeySchemaName() {
```
Comment (dasl-) [inline debezium-core/src/main/java/io/debezium/relational/topic/TopicMappingProvider.java]: Why are we providing a way to customize the KeySchemaName and ValueSchemaName? As I understand it, the motivation for this PR was custom DML topic mappings. That doesn't seem to require customizing the KeySchemaName or ValueSchemaName...

That said, I guess it doesn't hurt to add that customization feature.

**Comment by dasl-** [comment]
I'm excited about these changes. Let me know if I can help!

**Comment by rhauch** [comment]
@dasl- I hope to get back to this first. It is possible to limit the changes to only allow someone to simply customize the names of the topics, but not the table-to-topic mapping. This proposal tries to be even more flexible by allowing multiple tables to be mapping to a single topic, and to do this we'll likely need to augment the message keys since a table's primary key is no longer sufficient to distinguish the row in table A from perhaps a row from table B with the same primary key.

Having said all this, though, I've not really looked at this PR for some time, and I may want to do things differently. Any chance you want to take a whack at it, even if it's just a start/concept?

**Comment by dasl-** [comment]
@rhauch My approach would probably be very similar to the one you took here. I would like to allow multiple tables to map to a single topic also.

I will try applying this patch and seeing how it works. Perhaps I will have a follow up code review with some additions.

**Comment by rhauch** [comment]
Closed _without merging_. Instead, we've gone with using Single Message Transforms per #211.

**Pull request closed without merging**